Jumpscale/data/bcdb/BCDBMeta.py

```diff
@@ -43,103 +62,114 @@ def _reset_runtime_metadata(self):
-            s_obj.url = schema.url
-            s_obj.text = schema.text  # + "\n"  # only 1 \n at end
-            s_obj.md5 = schema._md5
-            s_obj.hasdata = schema.hasdata
-            self._url_mid_set(schema)
+        change = False  # we only want to save is there is a change
+
+        # deal with making sure that the md5 of this schema is registered as the newest one
+        if schema.url in self._data["url"]:
+            md5s = self._data["url"][schema.url]
+            if schema._md5 in md5s:
+                if schema._md5 != md5s[-1]:
+                    # means its not the latest one
+                    change = True
+                    md5s.pop(md5s.index(schema._md5))
+                    md5s.append(schema._md5)  # now at end of list again
+            else:
+                # is a new one, not in list yet
+                change = True
+                md5s = [schema._md5]
+        else:
+            change = True
+            md5s = [schema._md5]
+        if change:
+            self._data["url"][schema.url] = md5s
+
+        change2 = False
+        if schema._md5 not in self._data["md5"]:
+            change2 = True
+            d = {}
+            d["mid"] = bcdbmodel.mid
+            d["text"] = schema.text
+            d["hasdata"] = schema.hasdata
+            d["epoch"] = j.data.time.epoch
+            d["url"] = schema.url
+            self._data["md5"][schema._md5] = d
+
+        if change or change2:
             self._save()
 
-    def _add_to_schema_factory(self, schema_obj):
-        """
-        will make sure that everyone in kosmos has access to the schema's as found in the metadata of this bcdb
-        :param schema_obj:
-        :return:
-        """
-        s = schema_obj
-        isok = j.data.schema._add_url_to_md5(s.url, s.md5)
-        if not isok:
-            # means there is a newer one in memory
-            schema = j.data.schema.get_from_url(s.url)
-            self._schema_set(schema)  # tell our bdcdb metadata that this exists
-
         # don't load the full schema but put the text of schema there
-        if not j.data.schema.exists(md5=s.md5):
-            j.data.schema._md5_to_schema[s.md5] = s.text
-
-    def _url_mid_set(self, schema):
-        url = schema.url
-        last = 0
-        if not url in self._data.url_to_mid:
-            # look for the last one
-            for item in self._data.url_to_mid.values():
-                if int(item) > last:
-                    last = int(item)
-            self._data.url_to_mid[url] = last + 1
+        if not j.data.schema.exists(md5=schema._md5):
+            j.data.schema._md5_to_schema[schema._md5] = schema.text
+
+        # need to check in j.data.schema that this schema is the newest version
+        j.shell()
+
+    def _data_from_url(self, url):
+        if url not in self._data["url"]:
+            raise j.exceptions.Input("cannot find url in bcbd:%s" % url)
+        if len(self._data["url"][url]) == 0:
```

Comment:
if not self._data["url"][url]):

Reply:
not necessarely the same e.g. self._data["url"][url]=False would match ineed to check on []
its more specific, but when other value at least my code will barf, you'rs could continue, I do realize far fetched but thats why I try to avoid the generic not... statement in python